openssl-dynamic/src/main/c/ocsp.c

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "tcn.h"
+#include "ssl_private.h"
+
+// OCSP stapling should be present in OpenSSL as of version 1.0.0
+#ifndef TCN_OCSP_MIN_OPENSSL_VERSION_NUMBER
+#define TCN_OCSP_MIN_OPENSSL_VERSION_NUMBER 0x10002000L
+#endif /* TCN_OCSP_MIN_OPENSSL_VERSION_NUMBER */
+
+#ifndef TCN_NO_OCSP
+#if defined(OPENSSL_NO_OCSP) || (OPENSSL_VERSION_NUMBER < TCN_OCSP_MIN_OPENSSL_VERSION_NUMBER)
+#define TCN_NO_OCSP 1
+#endif
+#endif /* TCN_NO_OCSP */
+
+#if !defined(TCN_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL)
+
+static const int OCSP_CLIENT_ACK = 1;
+static const int OCSP_SERVER_ACK = SSL_TLSEXT_ERR_OK;
+
+/**
+ * This function is being called from OpenSSL. We do everything in
+ * Java-land and this callback is just a stub that returns the
+ * right values to keep OpenSSL happy.
+ *
+ * The arg that is passed into this function is the pointer for one
+ * of these values:
+ *   OCSP_CLIENT_ACK
+ *   OCSP_SERVER_ACK
+ */
+static int openssl_ocsp_callback(SSL *ssl, void *arg) {
+    return *(const int*)arg;
+}
+
+#endif /* !defined(TCN_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL) */
+
+/**
+ * Enables OCSP stapling for the given SSLContext
+ */
+TCN_IMPLEMENT_CALL(void, SSLContext, enableOcsp)(TCN_STDARGS, jlong ctx, jboolean client) {
+
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+    if (c == NULL) {
+        tcn_ThrowException(e, "ctx is null");
+        return;
+    }
+
+#if defined(TCN_NO_OCSP)
+    tcn_ThrowException(e, "OCSP stapling is not supported");
+    return;
+
+#elif defined(OPENSSL_IS_BORINGSSL)
+    // This does nothing in BoringSSL
+    return;
```

Comment:
Since this is a void method no need to explicitly return ... consider simplifying to:

```c
#if defined(TCN_NO_OCSP)
  // throw
#elif !defined(OPENSSL_IS_BORINGSSL) /* noop for BoringSSL */
 // do OpenSSL stuff
#endif
```